framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java

```diff
@@ -0,0 +1,2382 @@
+        processExports(javacTree, node);
+        visitLists(javacTree.getModuleNames(), node.getModuleNames());
+        javacTree.getPackageName().accept(this, node.getName());
+        return null;
+    }
+
+    @Override
+    public Void visitExpressionStatement(ExpressionStatementTree javacTree, Node javaParserNode) {
+        if (javaParserNode instanceof ExpressionStmt) {
+            ExpressionStmt node = (ExpressionStmt) javaParserNode;
+            processExpressionStatemen(javacTree, node);
+            javacTree.getExpression().accept(this, node.getExpression());
+        } else if (javaParserNode instanceof ExplicitConstructorInvocationStmt) {
+            // In this case the expression will be a MethodTree, which would be better to match with
+            // the statement than the expression statement itself.
+            javacTree.getExpression().accept(this, javaParserNode);
+        } else {
+            throwUnexpectedNodeType(javacTree, javaParserNode);
+        }
+
+        return null;
+    }
+
+    @Override
+    public Void visitForLoop(ForLoopTree javacTree, Node javaParserNode) {
+        if (!(javaParserNode instanceof ForStmt)) {
+            throwUnexpectedNodeType(javacTree, javaParserNode, ForStmt.class);
+        }
+
+        ForStmt node = (ForStmt) javaParserNode;
+        processForLoop(javacTree, node);
+        Iterator<? extends StatementTree> javacIter = javacTree.getInitializer().iterator();
+        for (Expression initializer : node.getInitialization()) {
+            if (initializer.isVariableDeclarationExpr()) {
+                for (VariableDeclarator declarator :
+                        initializer.asVariableDeclarationExpr().getVariables()) {
+                    assert javacIter.hasNext();
+                    javacIter.next().accept(this, declarator);
+                }
+            } else if (initializer.isAssignExpr()) {
+                ExpressionStatementTree statement = (ExpressionStatementTree) javacIter.next();
+                statement.getExpression().accept(this, initializer);
+            } else {
+                assert javacIter.hasNext();
+                javacIter.next().accept(this, initializer);
+            }
+        }
+
+        assert !javacIter.hasNext();
+        assert (javacTree.getCondition() != null) == node.getCompare().isPresent();
+        if (javacTree.getCondition() != null) {
+            javacTree.getCondition().accept(this, node.getCompare().get());
+        }
+
+        // Javac stores a list of expression statements and JavaParser stores a list of statements,
+        // the javac statements must be unwrapped.
+        assert javacTree.getUpdate().size() == node.getUpdate().size();
+        Iterator<Expression> javaParserIter = node.getUpdate().iterator();
+        for (ExpressionStatementTree update : javacTree.getUpdate()) {
+            update.getExpression().accept(this, javaParserIter.next());
```

Comment:
This seems to visit a javac expression and a JavaParser statement, which seems wrong to me because they are different kinds of AST nodes.  Perhaps clarifying the comment will resolve my confusion.